go.sum

```diff
@@ -1,6 +1,7 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
-github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+github.com/DataDog/zstd v1.4.1 h1:3oxKN3wbHibqx897utPC2LTQU4J+IHWWJO+glkAkpFM=
+github.com/DataDog/zstd v1.4.1/go.mod h1:1jcaCB/ufaK+sKp1NBhlGmpz41jOoPQ35bpF36t7BBo=
```

Comment:
If upgrading to newer Badger v3, there is a native Go version of zstd to [replace the datadog CGo dependency](https://github.com/dgraph-io/badger/pull/1709).

Caddy will still depend on the datadog one due to dependency on smallstep/nosql having a upgrade blocker to Badger v3, but a Caddy maintainer has set up a [PR for Badger v2 to backport this zstd dependency change](https://github.com/dgraph-io/badger/pull/1736).

Reply:
BTW, `go.sum` is not dependencies, it's just all the collected hashes of sources in the dependency graph. It doesn't actually mean that version will be used in the build.

> The go.sum file may contain hashes for multiple versions of a module. The go command may need to load go.mod files from multiple versions of a dependency in order to perform [minimal version selection](https://golang.org/ref/mod#minimal-version-selection).

Reply:
> BTW, `go.sum` is not dependencies, it's just all the collected hashes of sources in the dependency graph.

Thanks for the clarification! I'm a Go noob :grimacing: 

> It doesn't actually mean that version will be used in the build.

Is it building a dependency when I get errors like this from `xcaddy`?:

```
/go/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1-rc1.0.20201003150343-5d1bab4fc658/table/table.go:849:18: not enough arguments in call to z.Calloc   
```

Updating that RC version to newer using `xcaddy --with` resolved it. AFAIK that version of badger is/was a dependency of `smallstep/nosql` that was recently updated and Caddy will update as well. Presumably it was a minimal version for v2 because of the RC suffix, so semver rules to use a newer version were invalid? (_that or there was no other Badger v2 versions referenced so it settled on that one?_ :man_shrugging: )

The `go.sum` is like `yarn.lock`/`cargo.lock` from what you've described. Will it be stale and still reference the wrong version causing a build error after Caddy updates? (_assuming that the `go.sum` for this module will need to be updated, even if `go.mod` is updated or I use `xcaddy --with` for the newer Caddy release_)

Hopefully this PR is enough to prevent the build issues in future :sweat_smile:

Reply:
> Is it building a dependency when I get errors like this from xcaddy?:

`smallstep/nosql` just happened to have last updated its badger dependency to that `v2.0.1-rc1` release, so that's what was used. I don't know why that particular error was happening. Too deep down the chain, and I don't super care to dig to find out. I just want the CGO dependencies gone ASAP 😬 but the recent version bump to the latest v2 of badger should be a decent stopgap even if my PR for backporting the removal of `DataDog/zstd` isn't merged anytime soon.

> The go.sum is like yarn.lock/cargo.lock from what you've described.

That's essentially my understanding, yeah. Go stores the actual sources in your GOPATH somewhere, and those versions and hashes are references to specific versions, so Go can verify that what's in the `go.sum` matches what was downloaded out of band.

> Will it be stale and still reference the wrong version causing a build error after Caddy updates?

As the link I sent above explains, any `go get` or `go mod tidy` command updates the `go.sum` automatically, so it's typically kept up to date pretty well.